shared/extension/Behaviour.hpp

```diff
@@ -0,0 +1,317 @@
+                , name(name)
+                , priority(priority)
+                , optional(optional) {}
+
+            /// The provider type this director task is for
+            std::type_index type;
+            /// The reaction id of the requester (could be the id of a provider)
+            uint64_t requester_id;
+            /// The reaction task id of the requester (if it is a provider later a ProviderDone will be emitted)
+            uint64_t requester_task_id;
+            /// The data for the command, (the data that will be given to the provider)
+            std::shared_ptr<void> data;
+            /// A name for this task to be shown in debugging systems
+            std::string name;
+            /// The priority with which this task is to be executed
+            int priority;
+            /// If this task is optional
+            bool optional;
+        };
+
+        /**
+         * This type is used as a base type for the different provider DSL keywords (Provides, Entering, Leaving) to
+         * handle their common code.
+         *
+         * @tparam T        The type that this provider services
+         * @tparam action   The action type that this provider is
+         */
+        template <typename T, ProviderAction action>
+        struct ProviderBase {
+
+            template <typename DSL>
+            static inline void bind(const std::shared_ptr<NUClear::threading::Reaction>& reaction) {
+
+                // Tell the director
+                reaction->reactor.powerplant.emit(std::make_unique<ProvidesReaction>(reaction, typeid(T), action));
+
+                // Add our unbinder
+                reaction->unbinders.emplace_back([](const NUClear::threading::Reaction& r) {
+                    r.reactor.emit<NUClear::dsl::word::emit::Direct>(
+                        std::make_unique<NUClear::dsl::operation::Unbind<commands::ProvidesReaction>>(r.id));
+                });
+            }
+
+            template <typename DSL>
+            static inline std::shared_ptr<T> get(NUClear::threading::Reaction& r) {
+
+                // TODO make this return a special type that can be dereferenced into a T object
+                // TODO then you can add in extra information that can be passed to the provider (like when it is
+                // triggered due to a `Done` event)
+                // TODO here we need to get task information from the director
+                // TODO using r.id we can query the director if this provider is supposed to be running right now and
+                // also get the data for this provider
+                // If this provider is not supposed to be running right now we return a nullptr
+                // By returning nullptr that evaluates to something "falsy" it will prevent this reaction from running
+                // (unless the provider has an optional wrapper around it but then it's the users job to deal with that)
+                return nullptr;
```

Comment:
The comment begs the question - what's the reason against using `std::optional` here?

Reply:
This getter will actually have an entire type that's there for it so the user can either decide to have a full context from the Director or just a message that they care about.
I haven't written it yet as its contents will depend on what the algorithm ends up doing.